{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "606ae6b5",
   "metadata": {},
   "outputs": [],
   "source": [
    "from pysam import VariantFile"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "a45358d8",
   "metadata": {},
   "outputs": [],
   "source": [
    "childVCF = VariantFile(\"HG002-NA24385-50x_filtered.vcf\")\n",
    "fatherVCF = VariantFile(\"HG003.hs37d5.60x.1.converted_filtered.vcf\")\n",
    "motherVCF = VariantFile(\"HG004.hs37d5.60x.1.converted_filtered.vcf\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "4c24118c",
   "metadata": {},
   "outputs": [],
   "source": [
    "def findMaxChromPos(VCFs):\n",
    "    maxPos = 0\n",
    "    for vcf in VCFs:  \n",
    "        for rec in vcf.fetch():\n",
    "            maxPos = max(maxPos, rec.pos)\n",
    "    return maxPos, len(str(abs(maxPos)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "86454ae7",
   "metadata": {},
   "outputs": [],
   "source": [
    "#VCFs = (childVCF, fatherVCF, motherVCF)\n",
    "#count, digits = findMaxChromPos(VCFs)\n",
    "#print(\"Maximum position among all chromosomes: \" + str(count) + \", number of digits: \" + str(digits))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "d278b0e8",
   "metadata": {},
   "outputs": [],
   "source": [
    "# check if there's a variation which returns child to reference genome"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "48636434",
   "metadata": {},
   "outputs": [],
   "source": [
    "def checkPosIntersection(childVariant, parentVariant):  \n",
    "    \n",
    "    if childVariant.ref != parentVariant.ref:\n",
    "        return False, None\n",
    "    \n",
    "    childGT = next(childVariant.samples.itervalues())['GT']\n",
    "    parentGT = next(parentVariant.samples.itervalues())['GT']\n",
    "    gt01 = (0,1)\n",
    "    gt11 = (1,1)\n",
    "    gt12 = (1,2)\n",
    "    childALT = childVariant.alts\n",
    "    parentALT = parentVariant.alts\n",
    "    \n",
    "    if parentGT == gt01:\n",
    "        # parent has 0/1 genotype, child has 0/1 genotype\n",
    "        # if child and parent have the same REF and ALT -> return the variant\n",
    "        if childGT == gt01:\n",
    "            if childALT[0] == parentALT[0]:\n",
    "                return True, childVariant\n",
    "        # parent has 0/1 genotype, child has 1/1 genotype\n",
    "        # if child and parent have the same REF and ALT -> child variant with 0/1 gt goes into intersection\n",
    "        elif childGT == gt11:\n",
    "            if childALT[0] == parentALT[0]:\n",
    "                next(childVariant.samples.itervalues())['GT'] = (0,1)\n",
    "                return True, childVariant\n",
    "        # parent has 0/1 genotype, child has 1/2 genotype\n",
    "        # if one of child's alts is the same as parent's alt-> return the variant\n",
    "        else :\n",
    "            if childALT[0] == parentALT[0] or childALT[1] == parentALT[0]:\n",
    "                next(childVariant.samples.itervalues())['GT'] = (0,1)\n",
    "                childVariant.alts = parentVariant.alts\n",
    "                return True, childVariant\n",
    "    elif parentGT == gt11:\n",
    "        # parent has 1/1 genotype, child has 0/1 genotype\n",
    "        # if child's alt is the same as parent's alt -> return the variant\n",
    "        if childGT == gt01:\n",
    "            if childALT[0] == parentALT[0]:\n",
    "                return True, childVariant\n",
    "        # parent has 1/1 genotype, child has 1/1 genotype\n",
    "        # if child's alt is the same as parent's alt-> return the variant\n",
    "        elif childGT == gt11:\n",
    "            if childALT[0] == parentALT[0]:\n",
    "                return True, childVariant\n",
    "        # parent has 1/1 genotype, child has 1/2 genotype\n",
    "        # if one of child's alts is the same as parent's-> return the variant\n",
    "        else :\n",
    "            if childALT[0] == parentALT[0] or childALT[1] == parentALT[0]:\n",
    "                next(childVariant.samples.itervalues())['GT'] = (0,1)\n",
    "                childVariant.alts = parentVariant.alts\n",
    "                return True, childVariant\n",
    "    else :\n",
    "        # parent has 1/2 genotype, child has 0/1 genotype\n",
    "        # if child's alt is the same as one of parent's alts-> return the variant\n",
    "        if childGT == gt01:\n",
    "            if childALT[0] == parentALT[0] or childALT[0] == parentALT[1]:\n",
    "                next(childVariant.samples.itervalues())['GT'] = (0,1)\n",
    "                return True, childVariant\n",
    "        #parent has 1/2 genotype, child has 1/1 genotype\n",
    "        elif childGT == gt11:\n",
    "            if childALT[0] == parentALT[0] or childALT[0] == parentALT[1]:\n",
    "                next(childVariant.samples.itervalues())['GT'] = (0,1)\n",
    "                return True, childVariant\n",
    "        #parent has 1/2 genotype, child has 1/2 genotype\n",
    "        else :\n",
    "            #if child has both variations as parent\n",
    "            if childALT[0] == parentALT[0] and childALT[1] == parentALT[1] or childALT[0] == parentALT[1] and childALT[1] == parentALT[0]:\n",
    "                return True, childVariant\n",
    "            elif childALT[0] == parentALT[0] or childALT[0] == parentALT[1]:\n",
    "                next(childVariant.samples.itervalues())['GT'] = (0,1)\n",
    "                childVariant.alts = (childALT[0],)\n",
    "                return True, childVariant\n",
    "            elif childALT[1] == parentALT[0] or childALT[1] == parentALT[1]:\n",
    "                next(childVariant.samples.itervalues())['GT'] = (0,1)\n",
    "                childVariant.alts = (childALT[1],)\n",
    "                return True, childVariant\n",
    "            \n",
    "    return False, None"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "c87e063c",
   "metadata": {},
   "outputs": [],
   "source": [
    "def findIntersection(childVCF, parentVCF, outputFileName):\n",
    "    # check filename extension -----------------------\n",
    "    outputVCF = VariantFile(outputFileName, 'w', header=childVCF.header)\n",
    "    chrom = \"\"\n",
    "    for (childVariant, parentVariant) in zip(childVCF.fetch(),parentVCF.fetch()):\n",
    "        # sex chromosomes are skipped -------------------\n",
    "        if childVariant.chrom == \"chrX\" or parentVariant.chrom == \"chrX\": \n",
    "            return\n",
    "        #find the same variant position on the same chromosome for both\n",
    "        while childVariant.pos != parentVariant.pos or childVariant.chrom != parentVariant.chrom:\n",
    "            while childVariant.pos > parentVariant.pos:\n",
    "                if childVariant.chrom != parentVariant.chrom:\n",
    "                    childVariant = next(childVCF, None) \n",
    "                else:\n",
    "                    parentVariant = next(parentVCF, None)\n",
    "                if parentVariant is None or childVariant is None or parentVariant.chrom == \"chrX\" or childVariant.chrom == \"chrX\": \n",
    "                    return\n",
    "            while childVariant.pos < parentVariant.pos:\n",
    "                if childVariant.chrom != parentVariant.chrom:\n",
    "                    parentVariant = next(parentVCF, None)\n",
    "                else:\n",
    "                    childVariant = next(childVCF, None) \n",
    "                if parentVariant is None or childVariant is None or parentVariant.chrom == \"chrX\" or childVariant.chrom == \"chrX\": \n",
    "                    return\n",
    "        \n",
    "        #child variant position and parent variant position are matched\n",
    "        condition, variant = checkPosIntersection(childVariant, parentVariant)\n",
    "        if condition: \n",
    "            outputVCF.write(variant)\n",
    "    outputVCF.close()\n",
    "    #return outputVCF"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "4adf4f31",
   "metadata": {},
   "outputs": [],
   "source": [
    "#findIntersection(childVCF, motherVCF, \"outputCM.vcf\")\n",
    "#findIntersection(childVCF, fatherVCF, \"outputCF.vcf\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "f12ff25a",
   "metadata": {},
   "outputs": [],
   "source": [
    "def correctIntersectionToUnion(variant):\n",
    "    gt = next(variant.samples.itervalues())['GT']   \n",
    "    if gt == (1,1):\n",
    "        next(variant.samples.itervalues())['GT'] = (0,1)\n",
    "    elif gt == (1,2) :\n",
    "        next(variant.samples.itervalues())['GT'] = (0,1)\n",
    "        variant.alts = (variant.alts[0],)\n",
    "    return variant"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "eab660b7",
   "metadata": {},
   "outputs": [],
   "source": [
    "def findUnion(VCF1, VCF2, outputFileName):\n",
    "    chrom = \"\"\n",
    "    # check filename extension -----------------------\n",
    "    outputVCF = VariantFile(outputFileName, 'w', header=VCF1.header)\n",
    "    for (variant1, variant2) in zip(VCF1.fetch(),VCF2.fetch()):\n",
    "        #find the same variant position on the same chromosome for both\n",
    "        while variant1.pos != variant2.pos or variant1.chrom != variant2.chrom:\n",
    "            while variant1.pos > variant2.pos:\n",
    "                if variant1.chrom != variant2.chrom:\n",
    "                    #variant2 on that position is gt 0/0\n",
    "                    outputVCF.write(correctIntersectionToUnion(variant1)) \n",
    "                    variant1 = next(VCF1, None) \n",
    "                else:  \n",
    "                    #variant1 on that position is gt 0/0\n",
    "                    outputVCF.write(correctIntersectionToUnion(variant2))\n",
    "                    variant2 = next(VCF2, None)\n",
    "                if variant2 is None or variant1 is None: \n",
    "                    return\n",
    "            while variant1.pos < variant2.pos:\n",
    "                if variant1.chrom != variant2.chrom:\n",
    "                    #variant1 on that position is gt 0/0\n",
    "                    outputVCF.write(correctIntersectionToUnion(variant2))\n",
    "                    variant2 = next(VCF2, None)\n",
    "                else:\n",
    "                    #variant2 on that position is gt 0/0\n",
    "                    outputVCF.write(correctIntersectionToUnion(variant1))\n",
    "                    variant1 = next(VCF1, None) \n",
    "                if variant2 is None or variant1 is None : \n",
    "                    return\n",
    "        \n",
    "        #variant positions are matched\n",
    "        \n",
    "        gt1 = next(variant1.samples.itervalues())['GT']\n",
    "        gt2 = next(variant2.samples.itervalues())['GT']\n",
    "        \n",
    "        if gt1 == (0,1) and gt2 == (0,1):\n",
    "            if variant1.alts == variant2.alts:\n",
    "                next(variant1.samples.itervalues())['GT'] = (1,1)\n",
    "            else:\n",
    "                variant1.alts = (variant1.alts[0],variant2.alts[0])\n",
    "                next(variant1.samples.itervalues())['GT'] = (1,2)\n",
    "            outputVCF.write(variant1)\n",
    "        elif gt1 == (1,1) and gt2 == (1,1) or gt1 == (1,2) and gt2 == (1,2) or (gt2 == (0,1) and (gt1 == (1,1) or gt1 == (1,2))):\n",
    "            outputVCF.write(variant1)\n",
    "        elif gt1 == (0,1) and (gt2 == (1,1) or gt2 == (1,2)):\n",
    "            outputVCF.write(variant2)\n",
    "        else:\n",
    "            print(\"error \"+str(gt1)+\" \"+str(gt2))\n",
    "        \n",
    "        \n",
    "        if chrom != variant1.chrom:\n",
    "            chrom = variant1.chrom\n",
    "            print(\"union on \"+chrom)\n",
    "    \n",
    "    outputVCF.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "67031bae",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "union on chr1\n",
      "union on chr2\n",
      "union on chr3\n",
      "union on chr4\n",
      "union on chr5\n",
      "union on chr6\n",
      "union on chr7\n",
      "union on chr8\n",
      "union on chr9\n",
      "union on chr10\n",
      "union on chr11\n",
      "union on chr12\n",
      "union on chr13\n",
      "union on chr14\n",
      "union on chr15\n",
      "union on chr16\n",
      "union on chr17\n",
      "union on chr18\n",
      "union on chr19\n",
      "union on chr20\n",
      "union on chr21\n",
      "union on chr22\n"
     ]
    }
   ],
   "source": [
    "#findIntersection(childVCF, motherVCF, \"outputCM.vcf\")\n",
    "#findIntersection(childVCF, fatherVCF, \"outputCF.vcf\")\n",
    "outputCM = VariantFile(\"outputCM.vcf\")\n",
    "outputCF = VariantFile(\"outputCF.vcf\")\n",
    "findUnion(outputCM, outputCF, \"outputUnion.vcf\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "952aa5ca",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
