{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "606ae6b5",
   "metadata": {},
   "outputs": [],
   "source": [
    "from pysam import VariantFile"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a45358d8",
   "metadata": {},
   "outputs": [],
   "source": [
    "childVCF = VariantFile(\"HG002-NA24385-50x_filtered.vcf\")\n",
    "fatherVCF = VariantFile(\"HG003.hs37d5.60x.1.converted_filtered.vcf\")\n",
    "motherVCF = VariantFile(\"HG004.hs37d5.60x.1.converted_filtered.vcf\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4c24118c",
   "metadata": {},
   "outputs": [],
   "source": [
    "def findMaxChromPos(VCFs):\n",
    "    maxPos = 0\n",
    "    for vcf in VCFs:  \n",
    "        for rec in vcf.fetch():\n",
    "            maxPos = max(maxPos, rec.pos)\n",
    "    return maxPos, len(str(abs(maxPos)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "48636434",
   "metadata": {},
   "outputs": [],
   "source": [
    "def checkPosIntersection(childVariant, parentVariant):  \n",
    "    \n",
    "    if childVariant.ref != parentVariant.ref:\n",
    "        return False, None\n",
    "    \n",
    "    #for X or Y chromosomes, return child if ref, alts and pos is the same\n",
    "    if childVariant.chrom == \"chrX\" or childVariant.chrom == \"chrY\":\n",
    "        if childVariant.alts == parentVariant.alts:\n",
    "            return True, childVariant\n",
    "        else:\n",
    "            return False, None            \n",
    "    \n",
    "    childGT = next(childVariant.samples.itervalues())['GT']\n",
    "    parentGT = next(parentVariant.samples.itervalues())['GT']\n",
    "    gt01 = (0,1)\n",
    "    gt11 = (1,1)\n",
    "    gt12 = (1,2)\n",
    "    childALT = childVariant.alts\n",
    "    parentALT = parentVariant.alts\n",
    "    \n",
    "    if parentGT == gt01:\n",
    "        # parent has 0/1 genotype, child has 0/1 genotype\n",
    "        # if child and parent have the same REF and ALT -> return the variant\n",
    "        if childGT == gt01:\n",
    "            if childALT[0] == parentALT[0]:\n",
    "                return True, childVariant\n",
    "        # parent has 0/1 genotype, child has 1/1 genotype\n",
    "        # if child and parent have the same REF and ALT -> child variant with 0/1 gt goes into intersection\n",
    "        elif childGT == gt11:\n",
    "            if childALT[0] == parentALT[0]:\n",
    "                next(childVariant.samples.itervalues())['GT'] = (0,1)\n",
    "                return True, childVariant\n",
    "        # parent has 0/1 genotype, child has 1/2 genotype\n",
    "        # if one of child's alts is the same as parent's alt-> return the variant\n",
    "        else :\n",
    "            if childALT[0] == parentALT[0] or childALT[1] == parentALT[0]:\n",
    "                next(childVariant.samples.itervalues())['GT'] = (0,1)\n",
    "                childVariant.alts = parentVariant.alts\n",
    "                return True, childVariant\n",
    "    elif parentGT == gt11:\n",
    "        # parent has 1/1 genotype, child has 0/1 genotype\n",
    "        # if child's alt is the same as parent's alt -> return the variant\n",
    "        if childGT == gt01:\n",
    "            if childALT[0] == parentALT[0]:\n",
    "                return True, childVariant\n",
    "        # parent has 1/1 genotype, child has 1/1 genotype\n",
    "        # if child's alt is the same as parent's alt-> return the variant\n",
    "        elif childGT == gt11:\n",
    "            if childALT[0] == parentALT[0]:\n",
    "                return True, childVariant\n",
    "        # parent has 1/1 genotype, child has 1/2 genotype\n",
    "        # if one of child's alts is the same as parent's-> return the variant\n",
    "        else :\n",
    "            if childALT[0] == parentALT[0] or childALT[1] == parentALT[0]:\n",
    "                next(childVariant.samples.itervalues())['GT'] = (0,1)\n",
    "                childVariant.alts = parentVariant.alts\n",
    "                return True, childVariant\n",
    "    else :\n",
    "        # parent has 1/2 genotype, child has 0/1 genotype\n",
    "        # if child's alt is the same as one of parent's alts-> return the variant\n",
    "        if childGT == gt01:\n",
    "            if childALT[0] == parentALT[0] or childALT[0] == parentALT[1]:\n",
    "                next(childVariant.samples.itervalues())['GT'] = (0,1)\n",
    "                return True, childVariant\n",
    "        #parent has 1/2 genotype, child has 1/1 genotype\n",
    "        elif childGT == gt11:\n",
    "            if childALT[0] == parentALT[0] or childALT[0] == parentALT[1]:\n",
    "                next(childVariant.samples.itervalues())['GT'] = (0,1)\n",
    "                return True, childVariant\n",
    "        #parent has 1/2 genotype, child has 1/2 genotype\n",
    "        else :\n",
    "            #if child has both variations as parent\n",
    "            if childALT[0] == parentALT[0] and childALT[1] == parentALT[1] or childALT[0] == parentALT[1] and childALT[1] == parentALT[0]:\n",
    "                return True, childVariant\n",
    "            elif childALT[0] == parentALT[0] or childALT[0] == parentALT[1]:\n",
    "                next(childVariant.samples.itervalues())['GT'] = (0,1)\n",
    "                childVariant.alts = (childALT[0],)\n",
    "                return True, childVariant\n",
    "            elif childALT[1] == parentALT[0] or childALT[1] == parentALT[1]:\n",
    "                next(childVariant.samples.itervalues())['GT'] = (0,1)\n",
    "                childVariant.alts = (childALT[1],)\n",
    "                return True, childVariant\n",
    "            \n",
    "    return False, None"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a4f78d20",
   "metadata": {},
   "outputs": [],
   "source": [
    "def skipXYChrom(variant, vcf, useXY, isMother):\n",
    "    if variant is None:\n",
    "        return None\n",
    "    if variant.chrom == \"chrX\" or variant.chrom == \"chrY\": \n",
    "        if not useXY: \n",
    "            while variant.chrom != \"chrM\":\n",
    "                variant = next(vcf, None)\n",
    "            return variant\n",
    "        else: \n",
    "            if isMother:\n",
    "                while variant.chrom == \"chrY\":\n",
    "                    variant = next(vcf, None)\n",
    "                return variant\n",
    "            else:\n",
    "                while variant.chrom == \"chrX\":\n",
    "                    variant = next(vcf, None)\n",
    "                return variant  \n",
    "    else:\n",
    "        return variant"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c87e063c",
   "metadata": {},
   "outputs": [],
   "source": [
    "def findIntersection(childVCF, parentVCF, outputFileName, useXY, isMother, intersectionType, filterPass = True):\n",
    "    #useXY -> if true, intersection is done on all chromosomes, if false, X and Y are skipped\n",
    "    #isMother -> should be true if intersecting child with mother, false if intersecting child with father\n",
    "    #intersectionType -> if true, when intersecting with mother we skip Y, when intersecting with father we skip X,\n",
    "    # if false, we treat X and Y as any other chromosome\n",
    "    #filterPass -> if true, check intersection only for records with PASS in filter column, if false, check all\n",
    "    \n",
    "    # check filename extension -----------------------\n",
    "    outputVCF = VariantFile(outputFileName, 'w', header=childVCF.header)\n",
    "    chrom = \"\"\n",
    "    \n",
    "    for (childVariant, parentVariant) in zip(childVCF.fetch(),parentVCF.fetch()):\n",
    "        if intersectionType:            \n",
    "            childVariant = skipXYChrom(childVariant, childVCF, useXY, isMother)\n",
    "            parentVariant = skipXYChrom(parentVariant, parentVCF, useXY, isMother)\n",
    "        #find the same variant position on the same chromosome for both\n",
    "        while not (parentVariant is None or childVariant is None) and (childVariant.pos != parentVariant.pos or childVariant.chrom != parentVariant.chrom):\n",
    "            while not (parentVariant is None or childVariant is None) and (childVariant.pos > parentVariant.pos):\n",
    "                if childVariant.chrom != parentVariant.chrom:\n",
    "                    childVariant = next(childVCF, None)\n",
    "                    if intersectionType: childVariant = skipXYChrom(childVariant, childVCF, useXY, isMother)\n",
    "                else:\n",
    "                    parentVariant = next(parentVCF, None)\n",
    "                    if intersectionType: parentVariant = skipXYChrom(parentVariant, parentVCF, useXY, isMother)\n",
    "            while not (parentVariant is None or childVariant is None) and (childVariant.pos < parentVariant.pos):\n",
    "                if childVariant.chrom != parentVariant.chrom:\n",
    "                    parentVariant = next(parentVCF, None)\n",
    "                    if intersectionType: parentVariant = skipXYChrom(parentVariant, parentVCF, useXY, isMother)\n",
    "                else:\n",
    "                    childVariant = next(childVCF, None)\n",
    "                    if intersectionType: childVariant = skipXYChrom(childVariant, childVCF, useXY, isMother)\n",
    "        \n",
    "        #if one reaches end after chromosome M, iterate through the other\n",
    "        if parentVariant is None:\n",
    "            while not (childVariant is None):\n",
    "                childVariant = next(childVCF, None)\n",
    "        elif childVariant is None:\n",
    "            while not (parentVariant is None):\n",
    "                parentVariant = next(parentVCF, None)\n",
    "        #child variant position and parent variant position are matched\n",
    "        else:   \n",
    "            if filterPass and (next(childVariant.filter.iterkeys()) != 'PASS' or next(parentVariant.filter.iterkeys()) != 'PASS'): \n",
    "                continue\n",
    "           \n",
    "            condition, variant = checkPosIntersection(childVariant, parentVariant)\n",
    "            if condition: \n",
    "                outputVCF.write(variant)\n",
    "            \n",
    "            #printing\n",
    "            if chrom != childVariant.chrom:\n",
    "                chrom = childVariant.chrom\n",
    "                print(\"Intersection on \"+chrom)\n",
    "            \n",
    "    outputVCF.close()\n",
    "    #return outputVCF"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f12ff25a",
   "metadata": {},
   "outputs": [],
   "source": [
    "def correctIntersectionToUnion(variant):\n",
    "    #for X or Y chromosomes keep the 1/1 or 1/2 format\n",
    "    if variant.chrom == \"chrX\" or variant.chrom == \"chrY\":\n",
    "        return variant\n",
    "    \n",
    "    gt = next(variant.samples.itervalues())['GT']  \n",
    "    \n",
    "    #for somatic chromosomes, child can get only 1 chromsome from each parent, so reduce 1/1 or 1/2 to 0/1\n",
    "    if gt == (1,1):\n",
    "        next(variant.samples.itervalues())['GT'] = (0,1)\n",
    "    elif gt == (1,2) :\n",
    "        next(variant.samples.itervalues())['GT'] = (0,1)\n",
    "        variant.alts = (variant.alts[0],)\n",
    "        \n",
    "    return variant"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "eab660b7",
   "metadata": {},
   "outputs": [],
   "source": [
    "def findUnion(VCF1, VCF2, outputFileName):\n",
    "    #unionType -> same as intersectionType - if true, child-mother has only X and child-father only Y,\n",
    "    #if false, both mother and father have intersections on both X and Y\n",
    "    chrom = \"\"\n",
    "    # check filename extension -----------------------\n",
    "    outputVCF = VariantFile(outputFileName, 'w', header=VCF1.header)\n",
    "    for (variant1, variant2) in zip(VCF1.fetch(),VCF2.fetch()):\n",
    "        #find the same variant position on the same chromosome for both\n",
    "        while not (variant2 is None or variant1 is None) and (variant1.pos != variant2.pos or variant1.chrom != variant2.chrom):\n",
    "            while not (variant2 is None or variant1 is None) and (variant1.pos > variant2.pos):\n",
    "                if variant1.chrom != variant2.chrom:\n",
    "                    if variant1.chrom == \"chrY\" and variant2.chrom == \"chrX\":\n",
    "                        outputVCF.write(correctIntersectionToUnion(variant2)) \n",
    "                        variant2 = next(VCF2, None) \n",
    "                    else:\n",
    "                        #variant2 on that position is gt 0/0\n",
    "                        outputVCF.write(correctIntersectionToUnion(variant1)) \n",
    "                        variant1 = next(VCF1, None) \n",
    "                else:  \n",
    "                    #variant1 on that position is gt 0/0\n",
    "                    outputVCF.write(correctIntersectionToUnion(variant2))\n",
    "                    variant2 = next(VCF2, None)\n",
    "            while not (variant2 is None or variant1 is None) and (variant1.pos < variant2.pos):\n",
    "                if variant1.chrom != variant2.chrom:\n",
    "                    if variant1.chrom == \"chrX\" and variant2.chrom == \"chrY\":\n",
    "                        outputVCF.write(correctIntersectionToUnion(variant1)) \n",
    "                        variant1 = next(VCF1, None) \n",
    "                    else:\n",
    "                        #variant1 on that position is gt 0/0\n",
    "                        outputVCF.write(correctIntersectionToUnion(variant2))\n",
    "                        variant2 = next(VCF2, None)\n",
    "                else:\n",
    "                    #variant2 on that position is gt 0/0\n",
    "                    outputVCF.write(correctIntersectionToUnion(variant1))\n",
    "                    variant1 = next(VCF1, None) \n",
    "        \n",
    "        if variant1 is None:\n",
    "            while not (variant2 is None):\n",
    "                outputVCF.write(correctIntersectionToUnion(variant2))\n",
    "                variant2 = next(VCF2, None)\n",
    "        elif variant2 is None:\n",
    "            while not (variant1 is None):\n",
    "                outputVCF.write(correctIntersectionToUnion(variant1))\n",
    "                variant1 = next(VCF1, None)\n",
    "        else: \n",
    "            #variant positions are matched\n",
    "            \n",
    "            #printing\n",
    "            if chrom != variant1.chrom:\n",
    "                chrom = variant1.chrom\n",
    "                print(\"union on \"+chrom)\n",
    "                \n",
    "            #if position matched on X or Y chromosomes - that's the same (child's) variation\n",
    "            if variant1.chrom == \"chrX\" or variant1.chrom == \"chrY\":\n",
    "                outputVCF.write(variant1)\n",
    "                continue\n",
    "            \n",
    "            #if position matched on somatic chromosomes\n",
    "            gt1 = next(variant1.samples.itervalues())['GT']\n",
    "            gt2 = next(variant2.samples.itervalues())['GT']\n",
    "\n",
    "            if gt1 == (0,1) and gt2 == (0,1):\n",
    "                if variant1.alts == variant2.alts:\n",
    "                    next(variant1.samples.itervalues())['GT'] = (1,1)\n",
    "                else:\n",
    "                    variant1.alts = (variant1.alts[0],variant2.alts[0])\n",
    "                    next(variant1.samples.itervalues())['GT'] = (1,2)\n",
    "                outputVCF.write(variant1)\n",
    "            elif gt1 == (1,1) and gt2 == (1,1) or gt1 == (1,2) and gt2 == (1,2) or (gt2 == (0,1) and (gt1 == (1,1) or gt1 == (1,2))):\n",
    "                outputVCF.write(variant1)\n",
    "            elif gt1 == (0,1) and (gt2 == (1,1) or gt2 == (1,2)):\n",
    "                outputVCF.write(variant2)\n",
    "            else:\n",
    "                print(\"error \"+str(gt1)+\" \"+str(gt2))\n",
    "    \n",
    "    outputVCF.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8640e31e",
   "metadata": {},
   "outputs": [],
   "source": [
    "def findDisjunction(unionVCF, childVCF, outputFileName, filterPass = True):\n",
    "    chrom = \"\"\n",
    "    # check filename extension -----------------------\n",
    "    outputVCF = VariantFile(outputFileName, 'w', header=childVCF.header)\n",
    "    for (unionVariant, childVariant) in zip(unionVCF.fetch(),childVCF.fetch()):\n",
    "        #find the same variant position on the same chromosome for both\n",
    "        while not (childVariant is None or unionVariant is None) and (unionVariant.pos != childVariant.pos or unionVariant.chrom != childVariant.chrom):\n",
    "            while not (childVariant is None or unionVariant is None) and (unionVariant.pos > childVariant.pos):\n",
    "                if unionVariant.chrom != childVariant.chrom:\n",
    "                    #childVariant on that position is gt 0/0, but union exists\n",
    "                    print(\"error in disjunction on \"+str(unionVariant.pos))\n",
    "                    unionVariant = next(unionVCF, None) \n",
    "                else:  \n",
    "                    #unionVariant on that position is gt 0/0 -> De novo variation\n",
    "                    if filterPass and next(childVariant.filter.iterkeys()) == 'PASS':   \n",
    "                        outputVCF.write(childVariant)\n",
    "                    childVariant = next(childVCF, None)\n",
    "            while not (childVariant is None or unionVariant is None) and (unionVariant.pos < childVariant.pos):\n",
    "                if unionVariant.chrom != childVariant.chrom:\n",
    "                    #unionVariant on that position is gt 0/0 -> De novo variation\n",
    "                    if filterPass and next(childVariant.filter.iterkeys()) == 'PASS':   \n",
    "                        outputVCF.write(childVariant)\n",
    "                    childVariant = next(childVCF, None)\n",
    "                else:\n",
    "                    #childVariant on that position is gt 0/0, but union exists\n",
    "                    print(\"error in disjunction on \"+str(unionVariant.pos))\n",
    "                    unionVariant = next(unionVCF, None) \n",
    "        \n",
    "        \n",
    "        if unionVariant is None:\n",
    "            while not (childVariant is None):\n",
    "                outputVCF.write(childVariant)\n",
    "                childVariant = next(childVCF, None)\n",
    "        elif childVariant is None:\n",
    "            while not (unionVariant is None):\n",
    "                print(\"error in disjunction on \"+str(unionVariant.pos))\n",
    "                unionVariant = next(unionVCF, None)\n",
    "        else: \n",
    "            #variant positions are matched\n",
    "            unionGT = next(unionVariant.samples.itervalues())['GT']\n",
    "            childGT = next(childVariant.samples.itervalues())['GT']\n",
    "\n",
    "            if unionGT == (0,1) and (childGT == (1,1) or childGT == (1,2)) or unionGT == (1,1) and childGT == (1,2):\n",
    "                # de novo variation on one chromosome\n",
    "                if childGT == (1,2): childVariant.alts = (childVariant.alts[0],)\n",
    "                next(childVariant.samples.itervalues())['GT'] = (0,1)\n",
    "                outputVCF.write(childVariant)\n",
    "            #elif unionGT == childGT or unionGT == (1,1) and childGT == (0,1):\n",
    "            elif unionGT == (1,2) and (childGT == (0,1) or childGT == (1,1)):\n",
    "                print(\"error in disjunction on \"+str(childVariant.pos))\n",
    "\n",
    "            if chrom != unionVariant.chrom:\n",
    "                chrom = unionVariant.chrom\n",
    "                print(\"disjunction on \"+chrom)\n",
    "    \n",
    "    outputVCF.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "952aa5ca",
   "metadata": {},
   "outputs": [],
   "source": [
    "def deNovo():\n",
    "    findIntersection(childVCF, motherVCF, \"outputCM.vcf\", True, True, True)\n",
    "    findIntersection(childVCF, fatherVCF, \"outputCF.vcf\", True, False, True)\n",
    "    outputCM = VariantFile(\"outputCM.vcf\")\n",
    "    outputCF = VariantFile(\"outputCF.vcf\")\n",
    "    findUnion(outputCM, outputCF, \"outputUnion.vcf\")\n",
    "    outputUnion = VariantFile(\"outputUnion.vcf\")\n",
    "    findDisjunction(outputUnion, childVCF, \"outputDeNovo.vcf\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7e60adcb",
   "metadata": {},
   "outputs": [],
   "source": [
    "def deNovoPartial(outputNameCM, outputNameCF):\n",
    "    outputCM = VariantFile(outputNameCM)\n",
    "    outputCF = VariantFile(outputNameCF)\n",
    "    findUnion(outputCM, outputCF, \"outputUnion_Partial.vcf\")\n",
    "    outputUnion = VariantFile(\"outputUnion_Partial.vcf\")\n",
    "    findDisjunction(outputUnion, childVCF, \"outputDeNovo_Partial.vcf\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a577313f",
   "metadata": {},
   "outputs": [],
   "source": [
    "deNovo()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "2aa09e7e",
   "metadata": {},
   "outputs": [],
   "source": [
    "def analyzeResults(childFileName, deNovoFileNameProject, deNovoFileNamePartial, deNovoFileNameRTG):\n",
    "    childVCF = VariantFile(childFileName)\n",
    "    deNovoProjectVCF = VariantFile(deNovoFileNameProject)\n",
    "    deNovoPartialVCF = VariantFile(deNovoFileNamePartial)\n",
    "    deNovoRTGVCF = VariantFile(deNovoFileNameRTG)\n",
    "    countChild = 0\n",
    "    countChildFiltered = 0\n",
    "    for rec in childVCF.fetch():\n",
    "        countChild+=1\n",
    "        if next(rec.filter.iterkeys()) == 'PASS': countChildFiltered+=1\n",
    "    print(\"Total number of variants in child is \"+str(countChild))\n",
    "    print(\"Number of variants in child that passed filter is \"+str(countChildFiltered))\n",
    "    countDeNovo = 0\n",
    "    for rec in deNovoProjectVCF.fetch():\n",
    "        countDeNovo+=1\n",
    "    print(\"Number of de novo variants found by this project is \"+str(countDeNovo))\n",
    "    print(\"Percentage of de novo variants is \"+str(countDeNovo*100.0/countChildFiltered))\n",
    "    \n",
    "    countDeNovo = 0\n",
    "    for rec in deNovoPartialVCF.fetch():\n",
    "        countDeNovo+=1\n",
    "    print(\"Number of de novo variants found using RTG Tools intersection is \"+str(countDeNovo))\n",
    "    print(\"Percentage of de novo variants is \"+str(countDeNovo*100.0/countChildFiltered))\n",
    "    \n",
    "    countDeNovo = 0\n",
    "    for rec in deNovoRTGVCF.fetch():\n",
    "        countDeNovo+=1\n",
    "    print(\"Number of de novo variants found by RTG Tools VCFEval is \"+str(countDeNovo))\n",
    "    print(\"Percentage of de novo variants is \"+str(countDeNovo*100.0/countChildFiltered))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "90671bfd",
   "metadata": {},
   "outputs": [],
   "source": [
    "analyzeResults(\"/sbgenomics/project-files/HG002-NA24385-50x_filtered.vcf\", \"outputDeNovo.vcf\", \"outputDeNovo_Partial.vcf\", \"/sbgenomics/project-files/DeNovo_RTGToolsVCFEval.vcf\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "47de6a9b",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
